### PR TITLE
Add an `lhapdf_compatibility` module for LHAPDF

### PR DESCRIPTION
**Pull request by scarlehoff**
This branch is now an update on top of #1861 that adds the possibility of running fits using pdfflow instead of lhapdf (during the fit lhapdf is used only for the t0 covmat and that should be covered fine by pdfflow).

Instead of directly importing `lhapdf`, the `lhapdf` imports are contained in an `lhapdf_compatibility` module which, if `lhapdf` is installed will simply provide it normally and transparently so nothing changes.

_if_ `lhapdf` is not installed, it uses a combination of [lhadpf_management](https://github.com/scarlehoff/lhapdf_management) and [pdfflow](https://github.com/N3PDF/pdfflow) to do whatever is needed from `lhapdf`.

wrt to those two libraries, `lhapdf_management` is feature complete in that it does all the management part of LHAPDF (installing, updating, looking around the metadata of a PDF) but it does nothing with the PDF themselves.
`pdfflow` instead is a port of the interpolation and extrapolation parts of LHAPDF but it is not really being maintained anymore. For what we need in the fit is enough and python-only (so we can run fits normally) but I would not swap LHAPDF for this library.

However, by concentrating all that is needed from LHAPDF in a single module it will be easier to swap LHAPDF for some other library that might appear in the future. We can also more easily extend LHAPDF. This will be relevant for instance for the nuclear PDFs.

The way I've included the changes has a nice perk (/cc @AleCandido) which is the fact that one should be able to extract data/fktables from validphys without pdfflow or lhapdf (meaning, tensorflow is not needed for that).

### CODE REVIEW

**Comment by RoyStegeman** [comment]
Thanks, I'll have a look later but for now - do you want to keep `lhapdf_management` external to nnpdf?

**Comment by AleCandido** [comment]
It's great, thanks @scarlehoff!

I believe we might want to keep an optional dependency on LHAPDF as long as it will be used (so almost forever, from present perspective), to test the grids produced.
But this should not interfere with the fit, so the `lhapdf_compatibility` approach is great.

In particular: `lhapdf_compatibility` is greater for transitioning, but the moment we'd have another reliable replacement, whatever it is (if ever), we do not even need to keep it side-by-side with `lhapdf`.

**Comment by scarlehoff** [comment]
> Thanks, I'll have a look later but for now - do you want to keep lhapdf_management external to nnpdf?

Yes.

> I believe we might want to keep an optional dependency on LHAPDF as long as it will be used (so almost forever, from present perspective), to test the grids produced.

Yes, for now the dependency on LHAPDF is mandatory from conda (and so when using the official distribution of `nnpdf`). This PR just opens the possibility of not using it (i.e., in development environments).

In order to use this PR you need to install the code like `pip install .[nolha]`

File: validphys2/src/validphys/checks.py [hunk 1]
```diff
@@ -71,7 +70,7 @@ def check_can_save_grid(ns, **kwags):
     if not ns['installgrid']:
         return
 
-    write_path = lhapdf.paths()[-1]
+    write_path = lhaindex.get_lha_paths()[-1]
```
[the remark this answers is not in the document]
Comment (RoyStegeman) [reply validphys2/src/validphys/checks.py:73]: Or instead of using it, scrap that function everywhere. It doesn't improve readability since it's already very clear and just obfuscates which call is being made to lhapdf. I'm not a huge fan of functions like that tbh

File: validphys2/src/validphys/lhapdfset.py [hunk 2]
```diff
@@ -28,9 +28,10 @@
 """
 import logging
 
-import lhapdf
 import numpy as np
 
+from validphys.lhapdf_compatibility import make_pdf
```
Comment (RoyStegeman) [inline validphys2/src/validphys/lhapdfset.py:33]:
```suggestion
from validphys.lhapdf_compatibility as lhac
```
The more obvious `lc`  is already an existing variable in some places.

I'd prefer to keep the namespace a bit cleaner (especially if we import functions with such generic names as `make_pdf`).

File: validphys2/src/validphys/lhapdf_compatibility.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
+"""
+    Module for LHAPDF compatibility backends
+
+    If LHAPDF is installed, the module will transparently hand over everything to LHAPDF
+    if LHAPDF is not available, it will try to use a combination of the packages
+        `lhapdf-management` and `pdfflow`
+    which cover all the features of LHAPDF used during the fit (and likely most of validphys)
+
+    Eventually this module will allow us to transition to an under-development python/rust
+    PDF interpolation library.
```
Comment (RoyStegeman) [inline validphys2/src/validphys/lhapdf_compatibility.py:10]:
```suggestion
```

Or just comment on the possibility without making it specific for partons

File: validphys2/src/validphys/checks.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (scarlehoff) [reply validphys2/src/validphys/checks.py:73]: scrap lhapath you mean?

(I 100% agree with you on the obfuscation thing, but this might've been born many versions of lhapdf ago)

[the remark this answers is not in the document]
Comment (RoyStegeman) [reply validphys2/src/validphys/checks.py:73]: Scrap `get_lha_paths` and `get_lha_datapaths`. 

I genuinely don's see the benefit of having 
```
def get_lha_paths():
    return lhapdf.paths()


def get_lha_datapath():
    return get_lha_paths()[-1]
```
and now that `lhapdf` can be two different things it even hides that behind another layer, meaning I have to jump down to figure out if I'm dealing with `lhapdf_management` or `lhapdf`-classic.

[the remark this answers is not in the document]
Comment (scarlehoff) [reply validphys2/src/validphys/checks.py:73]: Fair. I don't even know why the two exist.
After the review I'll remove all instances and leave only one.

File: validphys2/src/validphys/lhio.py [hunk 4]
```diff
@@ -239,6 +232,7 @@ def new_pdf_from_indexes(
         files directly. It is slower and will call LHAPDF to fill the grids,
         but works for sets where the replicas have different grids.
     """
+    import lhapdf
```
[the remark this answers is not in the document]
Comment (scarlehoff) [reply validphys2/src/validphys/lhio.py:235]: it could, but I think you need lhapdf for that function (I hadn't really looked at it beyond hiding lhapdf so that it is not imported)

For some stuff you still need lhapdf, I wouldn't fully trust pdfflow.

File: validphys2/src/validphys/lhapdf_compatibility.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        res = self.pdf.py_xfxQ2_allpid(x, q**2).numpy()
+        return dict(zip(self._flavors, res.T))
+
+    def xfxQ(self, a, b, c=None):
+        """Wrapper for the LHAPDF xfxQ function
+        This is an overloaded function in LHAPDF so depending
+        on the number of arguments we will do:
+            xfxQ(flavours, x, Q)
+        or
+            xfxQ(x, q)
```
Comment (RoyStegeman) [inline validphys2/src/validphys/lhapdf_compatibility.py:75]: This makes me sad and confused. Why is this necessary?

**Review by RoyStegeman — COMMENTED** [review]
Everything looks okay. I'll have to take a bit of a leap of faith that pdfflow behaves the same as lhapdf and implements all the relevant functions, but the passing of the tests is a good sign. 

And since lhapdf is still the recommended backend to interact with PDFs it's not our responsibility if this breaks things (right??)

File: validphys2/src/validphys/lhapdf_compatibility.py (hunk 5 above)
Comment (scarlehoff) [reply validphys2/src/validphys/lhapdf_compatibility.py:75]: Because `lhapdf` passes this one to c++ under the hood where it is overloaded. `pdfflow` instead has two different functions. So it's a way of making it behave similarly.

**Comment by RoyStegeman** [comment]
What about postfit (and some other minor scripts that still use classic lhapdf)?

**Comment by RoyStegeman** [comment]
> The tests are always using the official lhapdf installation.

Haha I realized my mistake

**Review by RoyStegeman — APPROVED** [review]

**Comment by alecandido** [comment]
🎉
